lib_comfyui/webui_proxies.py

```diff
@@ -0,0 +1,317 @@
+            res = shared.sd_model.first_stage_model.encode(*args, **kwargs).cpu().share_memory_()
+            free_webui_memory()
+            return res
+
+    def decode(self, *args, **kwargs):
+        args = torch_utils.deep_to(args, device='cpu')
+        kwargs = torch_utils.deep_to(kwargs, device='cpu')
+        return torch_utils.deep_to(WebuiVaeProxy.sd_vae_decode(*args, **kwargs), device=self.device)
+
+    @ipc.confine_to('webui')
+    @staticmethod
+    def sd_vae_decode(*args, **kwargs):
+        args = torch_utils.deep_to(args, shared.sd_model.device)
+        kwargs = torch_utils.deep_to(kwargs, shared.sd_model.device)
+        with devices.autocast(), torch.no_grad():
+            res = shared.sd_model.first_stage_model.decode(*args, **kwargs).cpu().share_memory_()
+            free_webui_memory()
+            return res
+
+    def to(self, device):
+        assert str(device) == str(self.device), textwrap.dedent(f'''
+            cannot move the webui unet to a different device
+            comfyui attempted to move it from {self.device} to {device}
+        ''')
+        return self
+
+    def __getattr__(self, item):
+        if item in self.__dict__:
+            return self.__dict__[item]
+
+        res = WebuiVaeProxy.sd_vae_getattr(item)
+        if item != "device":
+            res = torch_utils.deep_to(res, device=self.device)
+
+        return res
+
+    @ipc.confine_to('webui')
+    @staticmethod
+    def sd_vae_getattr(item):
+        res = getattr(shared.sd_model.first_stage_model, item)
+        res = torch_utils.deep_to(res, 'cpu')
+        return res
+
+
+class WebuiClipWrapper:
+    def __init__(self, proxy):
+        self.cond_stage_model = proxy
+        self.patcher = WebuiModelPatcher(self.cond_stage_model)
+
+    def tokenize(self, *args, **kwargs):
+        args = torch_utils.deep_to(args, device='cpu')
+        kwargs = torch_utils.deep_to(kwargs, device='cpu')
+        return torch_utils.deep_to(WebuiClipWrapper.sd_clip_tokenize_with_weights(*args, **kwargs), device=self.cond_stage_model.device)
+
+    @ipc.confine_to('webui')
+    @staticmethod
+    def sd_clip_tokenize_with_weights(text, return_word_ids=False):
+        chunks, tokens_count, *_ = shared.sd_model.cond_stage_model.tokenize_line(text)
+        weighted_tokens = [list(zip(chunk.tokens, chunk.multipliers)) for chunk in chunks]
+        clip_max_len = shared.sd_model.cond_stage_model.wrapped.max_length
+        if return_word_ids:
+            padding_tokens_count = ((tokens_count // clip_max_len) + 1) * clip_max_len
+            for token_i in range(padding_tokens_count):
+                actual_id = token_i if token_i < tokens_count else 0
+                weighted_tokens[token_i // clip_max_len][token_i % clip_max_len] += (actual_id,)
+
+        return weighted_tokens
+
+    @property
+    def layer_idx(self):
+        clip_skip = self.cond_stage_model.config.num_hidden_layers - webui_settings.opts.CLIP_stop_at_last_layers
```

Comment:
are you sure this is fine?

Reply:
thanks, now it should be implemented with respect to the clip skip node in comfy